### PR TITLE
Mirror Projection

### PR DESCRIPTION
**Pull request by dmdashenkov**
## Introducing `Mirror` Projections.

A mirror projection is a system projection which copies the state of a domain aggregate. The projection then can be queried more efficiently.

In order for an entity to have a `Mirror`, the entity state should be marked as `AGGREGATE`:
```proto
message Order {
    option (entity).kind = AGGREGATE;
    
    // ...
}
```

Any query for an aggregate state is now dispatched by the `QueryService` to the `Mirror` projection repository.

An aggregate state can be queried:
 1. by ID(s);
 2. by the lifecycle flags;
 3. for all the states.

The combinations of the above query types work as well.

As the result of changing the query mechanism, the `StandStorage` has been deleted.

## Out of Scope

There are at least two major improvements TBD:
 1. Allowing not to post `EntityStateChanged` events for certain entities or cases.
 2. Allowing to query aggregate states by entity columns.

### `@Column` Contract

The entity columns cannot physically be extracted from a non-`public` class. Thus, a corresponding note was added to the `@Column` annotation doc.

### `ShardingReset`

A JUnit extension called `ShardingReset` is added to the server test library. The extension resets the sharding settings before _and_ after each test case.

To use the extension, configure it for a test suite with `@ExtendWith(ShardingReset.class)`.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/server/BoundedContext.java [hunk 1]
```diff
@@ -311,6 +311,20 @@ public CommandErrorHandler createCommandErrorHandler() {
...
+     * an {@link IllegalStateException}.
+     *
+     * <p>If a repository is registered, the method returns it or {@link Optional#empty()} if
+     * the requested entity is {@linkplain Visibility#NONE not visible}.
+     *
+     * @param entityStateClass
+     *         the class of the state of the entity managed by the resulting repository
+     * @return repository or {@link Optional#empty()} if the requested entity is not visible
+     * @throws IllegalStateException
+     *         if the requester repository is not registered
```
Comment (armiol) [inline server/src/main/java/io/spine/server/BoundedContext.java]: `requested`.

File: server/src/main/java/io/spine/server/BoundedContext.java [hunk 2]
```diff
@@ -311,6 +311,20 @@ public CommandErrorHandler createCommandErrorHandler() {
...
+     * <p>This method assumes that a repository for the given entity state class <b>is</b>
+     * registered in this context. If there is no such repository, throws
+     * an {@link IllegalStateException}.
+     *
+     * <p>If a repository is registered, the method returns it or {@link Optional#empty()} if
+     * the requested entity is {@linkplain Visibility#NONE not visible}.
+     *
+     * @param entityStateClass
+     *         the class of the state of the entity managed by the resulting repository
+     * @return repository or {@link Optional#empty()} if the requested entity is not visible
```
Comment (armiol) [inline server/src/main/java/io/spine/server/BoundedContext.java]: That's not an entity which isn't visible.

File: server/src/main/java/io/spine/server/stand/Stand.java [hunk 3]
```diff
@@ -440,31 +384,24 @@ public void close() throws Exception {
...
-            result = NOOP_PROCESSOR;
+            return NOOP_PROCESSOR;
         }
-        return result;
     }
 
-    private StandStorage getStorage() {
-        checkState(storage != null, "Stand %s does not have a storage assigned", this);
-        return storage;
+    private static QueryProcessor genericEntityProcessor(RecordBasedRepository<?, ?, ?> repository) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/stand/Stand.java]: I think we can safely inline this method.

File: server/src/main/java/io/spine/system/server/DefaultSystemGateway.java [hunk 4]
```diff
@@ -79,6 +83,21 @@ public void postEvent(Message systemEvent) {
               .post(event, noOpObserver());
     }
 
+    @Override
+    public Iterator<Any> readDomainAggregate(Query query) {
+        @SuppressWarnings("unchecked") // Logically checked.
+        MirrorRepository repository = (MirrorRepository)
+                system.findRepository(Mirror.class)
+                      .orElseThrow(
+                              () -> newIllegalStateException(
```
Comment (armiol) [inline server/src/main/java/io/spine/system/server/DefaultSystemGateway.java]: Let's discuss (in person!) if this is a proper behavior in a context of propagating this exception to the client.

File: server/src/main/java/io/spine/system/server/MirrorProjection.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    @Subscribe
+    public void on(EntityStateChanged event) {
+        getBuilder().setId(getId())
+                    .setState(event.getNewState());
+    }
+
+    @Subscribe
+    public void on(@SuppressWarnings("unused")
```
Comment (armiol) [inline server/src/main/java/io/spine/system/server/MirrorProjection.java]: Please have a comment on suppressions. Perhaps, in the `@implNote` section on a class-level.

File: server/src/main/java/io/spine/system/server/SystemGateway.java [hunk 6]
```diff
@@ -57,4 +61,23 @@
     static SystemGateway newInstance(SystemContext system) {
         return new DefaultSystemGateway(system);
     }
+
+    /**
+     * Executes the given query for a domain aggregate state.
+     *
+     * <p>This read operation supports following types of queries:
```
Comment (armiol) [inline server/src/main/java/io/spine/system/server/SystemGateway.java]: Not sure how to query by a `@Column` value. Let's discuss in person.

File: server/src/main/proto/spine/system/server/q/mirror.proto [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    option (entity).kind = PROJECTION;
+
+    MirrorId id = 1 [(required) = true, (valid) = true];
+
+    // The state of the aggregate.
+    google.protobuf.Any state = 2 [(required) = true, (valid) = true];
+
+    // The lifecycle flags of the aggregate.
+    //
+    // This field is here for conveniance. The own lifecycle flags of a mirror projection duplicate
```
Comment (armiol) [inline server/src/main/proto/spine/system/server/q/mirror.proto]: `convenience`

File: server/src/main/proto/spine/system/server/q/mirror.proto [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+message MirrorId {
+
+    // The ID of the corresponding aggregate packed as Any.
+    google.protobuf.Any value = 1 [(required) = true, (valid) = true];
+}
+
+// Entity columns of the associated aggregate.
+message EntityColumns {
+
+    // TODO:2018-09-06:dmytro.dashenkov: Support entity columns.
```
Comment (armiol) [inline server/src/main/proto/spine/system/server/q/mirror.proto]: Please have an issue and a link here.

Also, with the message of this structure (no reserved fields, for instance) I don't see how you are going to implement such a support.

File: server/src/test/java/io/spine/system/server/MirrorEventTests.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+import static io.spine.testing.server.projection.ProjectionEventDispatcher.dispatch;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * @author Dmytro Dashenkov
+ */
+@DisplayName("Mirror projection should on")
+class MirrorEventTests {
```
Comment (armiol) [inline server/src/test/java/io/spine/system/server/MirrorEventTests.java]: It looks `MirrorEventsTest`.

**Comment by dmdashenkov** [comment]
@armiol, let's discuss the leftover comments in person.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/system/server/MirrorRepository.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ * <p>In other cases, an entity won't have a {@link Mirror}.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class MirrorRepository
+        extends SystemProjectionRepository<MirrorId, MirrorProjection, Mirror> {
+
+    // TODO:2018-09-06:dmytro.dashenkov: Use for querying projection states.
+    // todo            https://github.co/SpineEventEngine/core-java/issues/840
```
Comment (armiol) [inline server/src/main/java/io/spine/system/server/MirrorRepository.java]: The link is broken.

File: server/src/test/java/io/spine/system/server/DefaultSystemGatewayTest.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                    .newBuilder()
+                    .setId(aggregateId)
+                    .build();
+            Command cmd = actorRequestFactory.createCommand(command);
+            domainContext.getCommandBus()
+                         .post(cmd, noOpObserver());
+        }
+    }
+
+    private ShoppingList findAggreagte(BoundedContext context) {
```
Comment (armiol) [inline server/src/test/java/io/spine/system/server/DefaultSystemGatewayTest.java]: Please fix a typo.

File: server/src/test/java/io/spine/system/server/MemoizingGateway.java [hunk 12]
```diff
@@ -139,4 +152,13 @@ public TenantId tenant() {
     public @Nullable MemoizedMessage lastSeenEvent() {
         return lastSeenEvent;
     }
+
+    /**
+     * Obtains the last query submitted to {@link SystemGateway}.
+     *
+     * @return {@code null} if no queries were submitted yet
+     */
+    public @Nullable MemoizedMessage lastSeenQuery() {
```
Comment (armiol) [inline server/src/test/java/io/spine/system/server/MemoizingGateway.java]: Let's use `Optional` here.

File: server/src/test/proto/spine/test/system/server/mirror_test.proto [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    PhotoId title_shot = 2 [(valid) = true];
+
+    spine.net.Url file_url = 3 [(valid) = true];
+
+    spine.net.Url sound_track_url = 4 [(valid) = true, (required) = false];
+
+    spine.net.Url subtitles_url = 5 [(valid) = true, (required) = false];
+}
+
+// This aggregate type is not finished, so the `(entity)` option is absent.
```
Comment (armiol) [inline server/src/test/proto/spine/test/system/server/mirror_test.proto]: Please explain what you wanted to say by "not finished" in more details. 

As it is now, it's hardly usable by anyone else except for the author.

**Pull request merged**